src/Microsoft.AspNetCore.Server.IntegrationTesting/TestApplication.cs

```diff
@@ -0,0 +1,215 @@
+
+                var hostProcess = new Process() { StartInfo = startInfo };
+
+                logger.LogInformation($"Executing command {DotnetCommandName} {parameters}");
+
+                hostProcess.StartAndCaptureOutAndErrToLogger("dotnet-publish", logger);
+
+                // A timeout is passed to Process.WaitForExit() for two reasons:
+                // 
+                // 1. When process output is read asynchronously, WaitForExit() without a timeout blocks until child processes
+                //    are killed, which can cause hangs due to MSBuild NodeReuse child processes started by dotnet.exe.
+                //    With a timeout, WaitForExit() returns when the parent process is killed and ignores child processes.
+                //    https://stackoverflow.com/a/37983587/102052
+                // 
+                // 2. If "dotnet publish" does hang indefinitely for some reason, tests should fail fast with an error message.
+                const int timeoutMinutes = 5;
+                if (hostProcess.WaitForExit(milliseconds: timeoutMinutes * 60 * 1000))
+                {
+                    if (hostProcess.ExitCode != 0)
+                    {
+                        var message = $"{DotnetCommandName} publish exited with exit code : {hostProcess.ExitCode}";
+                        logger.LogError(message);
+                        throw new Exception(message);
+                    }
+                }
+                else
+                {
+                    var message = $"{DotnetCommandName} publish failed to exit after {timeoutMinutes} minutes";
+                    logger.LogError(message);
+                    throw new Exception(message);
+                }
+
+                logger.LogInformation($"{DotnetCommandName} publish finished with exit code : {hostProcess.ExitCode}");
+            }
+        }
+
+        private string CopyPublishedOutput(string path, ILogger logger)
+        {
+            var target = CreateTempDirectory();
+
+            var source = new DirectoryInfo(path);
+            CopyFiles(source, target, logger);
+            return target.FullName;
+        }
+
+        private static DirectoryInfo CreateTempDirectory()
+        {
+            var tempPath = Path.GetTempPath() + Guid.NewGuid().ToString("N");
+            var target = new DirectoryInfo(tempPath);
+            target.Create();
+            return target;
+        }
+
+        public static void CopyFiles(DirectoryInfo source, DirectoryInfo target, ILogger logger)
+        {
+            foreach (DirectoryInfo directoryInfo in source.GetDirectories())
+            {
+                CopyFiles(directoryInfo, target.CreateSubdirectory(directoryInfo.Name), logger);
+            }
+
+            logger.LogDebug($"Processing {target.FullName}");
+            foreach (FileInfo fileInfo in source.GetFiles())
+            {
+                logger.LogDebug($"  Copying {fileInfo.Name}");
```

Comment:
This log seem too frequent.

Reply:
I thought so too. But it allowed me to see that we publish "refs" directory every time. So maybe it's not that bad?

Reply:
Maybe just print directories added?